src/Components/Components/src/NavigationManager.cs

```diff
@@ -262,6 +269,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
+
+                if (context.IsCanceled || cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
+        {
+            // If it was our cancellation token, we should gracefully cancel the navigation.
+            return true;
+        }
+        catch (Exception ex)
+        {
+            throw new LocationChangeException("An exception occurred while dispatching a location changing event.", ex);
+        }
+        finally
+        {
+            ArrayPool<IHandleLocationChanging>.Shared.Return(locationChangingHandlersCopy);
+        }
+    }
+
+    /// <summary>
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
+    protected virtual bool SetHasLocationChangingHandlers(bool value)
+        => true;
+
+    /// <summary>
+    /// Invokes <see cref="SetHasLocationChangingHandlers(bool)"/> if the existence of location changing handlers has changed.
+    /// </summary>
+    protected void UpdateHasLocationChangingHandlers()
+    {
+        var hasLocationChangingHandlers = _locationChangingHandlers.Count != 0;
+        if (hasLocationChangingHandlers != _lastUpdatedHasLocationChangingHandlers)
+        {
+            if (SetHasLocationChangingHandlers(hasLocationChangingHandlers))
+            {
+                _lastUpdatedHasLocationChangingHandlers = hasLocationChangingHandlers;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Adds a handler to process incoming navigation events.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to process incoming navigation events.</param>
+    public void AddLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+        _locationChangingHandlers.Add(locationChangingHandler);
+        UpdateHasLocationChangingHandlers();
+    }
+
+    /// <summary>
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
+    public bool RemoveLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+
+        if (_locationChangingHandlers.Remove(locationChangingHandler))
+        {
+            UpdateHasLocationChangingHandlers();
+            return true;
+        }
+
+        return false;
+    }
```

Comment:
Is there a reason why this is not an `event`?

Reply:
Is it because the callbacks are async?

Reply:
@javiercn Yes, and also having an event it makes it easier to accidentally register an anonymous function as a `LocationChanging` handler and never unregister it.

Reply:
@MackinnonBuck I can accept the first argument, but the second one is not really a valid concern. People is responsible for registering/unregistering events, the framework should not do anything about it.

I think we shouldn't add the interface, we should just take a `Func<LocationChangingContext, ValueTask>` in the methods. Otherwise, it makes it cumbersome to use the API.

Reply:
@javiercn That sounds good to me. FWIW, there was a discussion on this topic previously: https://github.com/dotnet/aspnetcore/pull/24417#discussion_r512163217